### ISSUE TITLE
Error: Ice connection failed

### ISSUE THREAD

**Issue by doganmurat** [feross/simple-peer#202]
Hi,
I am trying to make a connection with simple-peer.
One side is on node & other side is on chrome.
Normally everything is OK.
But When one side is switch to phone internet I get "Error: Ice connection failed"  error.
I have a  turn server & using  for iceServer config.
Any idea how to fix or debug that?

Thanks.

**Comment by vblagomir** [comment]
+1, have the same issue

**Comment by kifhan** [comment]
Have you try reconnectTimer option?
It attempt reconnect before emitting 'close'.

**Comment by vblagomir** [comment]
Have tried the reconnectTimer (set to 3000ms) it just tries one more time to reconnect (as expected) and still the same error at the end. As well tried in Safari's 'Develop > WebRTC' menu 'Disable ICE candidates restrictions' - the same error.

Btw, the signalling data exposed and sent (over separate websockets connection) contains real IP address of the peer on the mobile device, and is successfully received on desktop peer. So it might be not the issue with the ICE Candidates Restrictions.

Tried Safari 11 desktop and on iOS 11.

**Comment by vblagomir** [comment]
I have tried a custom STUN/TURN server and managed to fix the error.

**Comment by doganmurat** [comment]
Here is what I have tried;
I am using phone's internet sharing on my pc;

With an Android phone I could connect over relay server but with an iPhone I could not get a connection.
Then I forced to work with relay with iceTransportPolicy: "relay" option.
Now with iPhone I can connect but can not connect over normal network.

Also tried reconnectTimer:8000
Nothing changed.

Any idea?

**Comment by fabioaccetta** [comment]
@vblagomir

**Comment by vblagomir** [comment]
@fabioaccetta As a test I have created a stun/turn server here: http://numb.viagenie.ca and it worked.. now figuring out how to run my own..

**Comment by doganmurat** [comment]
@vblagomir Did you tried to connect by using iPhone (Shared Connection)?

**Comment by vblagomir** [comment]
So far I have entered following into ice servers list:
```
   {
      "urls": "stun:numb.viagenie.ca",
      "username": “some@email.com",
      "credential": “somepassword”
    },
    {
      "urls": "turn:numb.viagenie.ca",
      "username": "some@email.com",
      "credential": "somepassword"
    }
```
and managed to establish a data channel between iPhone iOS11 (connected in 3G network) and my local computer with Safari 11.0. But this is temporary and sort of proof of concept. Normally one should set up his/her own Turn/Stun server.

**Comment by MaciejWWojcik** [comment]
I have the same problem as You @doganmurat  a few days ago.
Finally I have repair this, and the config of simplePeer that works in my app is this:

`const peer = new SimplePeer({
      initiator: true,
      stream: this.videoStream,
      reconnectTimer: 100,
      iceTransportPolicy: 'relay',
      trickle: false,
      config: {
        iceServers: [
          {
            "urls": "stun:numb.viagenie.ca",
            "username": '******@gmail.com',
            "credential": '****'
          },
          {
            "urls": "turn:numb.viagenie.ca",
            "username": "****@gmail.com",
            "credential": "*******"
          }
        ]
      }
    });`

**Comment by doganmurat** [comment]
Hi @MaciejWWojcik ,

Please see my comment above.
If you force simple-peer to use relay server, then somehow you can not connect over normal network.
Could you please try this as well?

**Comment by ibudisteanu** [comment]
I can not make the Simple Peer to connect. It shows me the above-mentioned error. Is it because we use Simple-Peer? I tried all the above suggestions, and absolutely none of them worked for me. I tried dozens of iceServers. Any suggestions of how you made it work?  You can check my application on my [website](http://webdollar.io)

The error I receive

```
error Error: Ice connection failed.
    at Peer._onIceStateChange
```

**Comment by ibudisteanu** [comment]
I see that computers from  **some networks connect**, while other networks can't connect, although I have provided both STUN & TURN ICE servers. **Very probably the simple peers library is not trying to use the 2nd ICE server namely my  TURN ICE server**

![image](https://user-images.githubusercontent.com/17750489/32367782-0aeb9728-c08c-11e7-9fad-a8bdf4936231.png)

```
params = {
    initiator: initiator,
    trickle: false,
    reconnectTimer: 100,
    iceTransportPolicy: 'relay',
    config: {

        iceServers: [
            {
                urls: "stun:numb.viagenie.ca",
                username: "pasaseh@ether123.net",
                credential: "12345678"
            },
            {
                urls: "turn:numb.viagenie.ca",
                username: "pasaseh@ether123.net",
                credential: "12345678"
            }
        ]
    }
};
```

**Comment by RationalCoding** [comment]
I've only seen this issue when one or both peers are behind a symmetric NAT, so the problem does lie with the connection to the TURN server. Has anyone tried a non-viagenie TURN server so that we can rule out them being misconfigured?

Could you also confirm you are behind a symmetric NAT by running this JSFiddle on both peers? (https://jsfiddle.net/5ftsd5c2/17/)

**Comment by ibudisteanu** [comment]
Hi @RationalCoding  thanks for helping us out. On the network that other peers can connect the script shows "normal nat", while on the computer that can't connect and can't receive any connections shows "symmetric nat". I suspect the TURN server is never used.

**Comment by RationalCoding** [comment]
I'm noticing there are some reports of viagenie servers not properly collecting relay entries on some browsers, which could explain this. I'll try to set up a coturn server this weekend (or if someone has access to one, please message me).

The ice failures without TURN are due to symmetric NATs, there's nothing to do to fix that.

This isn't really related to simple-peer, but we get so many of these issues I'll leave it open.

**Comment by ibudisteanu** [comment]
@RationalCoding  maybe this can help you out to figure the problem. I can run this demo https://www.webrtc-experiment.com/text-chat/#4123528045277047 connecting the computers fast and without any problem.

**Comment by RationalCoding** [comment]
@ibudisteanu That demo doesn't seem to work at all. (Content-security errors and the signal server seems down.)

**Comment by RationalCoding** [comment]
Here's what I've found so far:

- The problem only occurs when one peer has a symmetric NAT and the other peer has any kind of NAT (symmetric or not).
- It's not related to the setup of the STUN or TURN servers (unless they are somehow all misconfigured, which is unlikely. Coturn and viagenie have the same behaviour.)
- Trickle, reconnectTimer and iceTransportPolicy don't affect it.
- Any combination of browsers will throw the error.

In Chrome logging, the error is this:
```
[39692:35211:1103/183902.705873:ERROR:sctptransport.cc(789)] SctpTransport->SendQueuedStreamResets(): Failed to send a stream reset for 1 streams: [0x00000009] Bad file descriptor
```

I haven't been able to find out if Firefox throws a similar error.

**Comment by ibudisteanu** [comment]
@RationalCoding thanks for your time and nice finding. But, how can we solve this problem? I can confirm that I tried setting different values for trickle and iceTransportPolicy and similar effect.

I see that now this [demo](https://www.webrtc-experiment.com/text-chat/#4123528045277047) no longer works, but yesterday it worked ( i have tested it many times... doh' on Unix and Windows). Now the signaling server is down.

This Web App definitely works and it is online https://opentokrtc.com/ and I have tested it right now on this kind of specific network ( Symmetric NAT and normal NAT) and it works, while the Simple Peer library doesn't work on this kind of connections.

**Comment by RationalCoding** [comment]
@ibudisteanu I don't know a solution yet.

In your website's case, in seems your TURN server isn't working at all. Try [this tool](https://webrtc.github.io/samples/src/content/peerconnection/trickle-ice/) with only your TURN server and it'll come up as unreachable. There should be several candidates there.

**Comment by ibudisteanu** [comment]
@RationalCoding  wait, I don't get it. You are just saying that the turn server (relay option) turn:numb.viagenie.ca is offline? So, it means that the problem is just to choose a different TURN server?

I see that these are TURN servers (relay shows available)?

stun:stun.l.google.com:19302

{  //quite slow
    url: 'turn:192.158.29.39:3478?transport=tcp',
    credential: 'JZEOEt2V3Qb0y27GRntt2u2PAYA=',
    username: '28224511:1379330808'
},

{//quite slow
    url: 'turn:turn.anyfirewall.com:443?transport=tcp',
    credential: 'webrtc',
    username: 'webrtc'
}

Did you try them?

**Comment by RationalCoding** [comment]
They both work. turn:numb.viagenie.ca is working fine now, but was unreachable before. It's ok, you don't have two problems.

**Comment by ibudisteanu** [comment]
I tried the turn:numb.viagenie.ca  and it looks unreachable. See the attached photo

![image](https://user-images.githubusercontent.com/17750489/32400082-8db6d38e-c104-11e7-93f2-9a3ad0cc2c8b.png)

I am new to WebRTC, you are just saying that I should change the offline TURN server (turn:numb.viagenie.ca)  to one that is reachable and it should work. Right?

**Comment by RationalCoding** [comment]
If your TURN server is going doing frequently, that's going to be a problem. I would ask viagenie for support or use a different server.

There's another problem with TURN servers in general, which is what this issue is about.

**Comment by ibudisteanu** [comment]
I guessed the issue reported is that we can't connect p2p via symmetric NAT using Simple Peer.  Because my ICE servers list is using the viagenie for TURN, and their turn servers are not reachable at my end, the Simple Peer tries to connect the symmetric NAT only via the STUN servers (which is impossible).    Here are my questions.

Can I use a longer TURN list to include a few more turn servers? 
Will the library automatically try to establish the p2p connection using all the TURN servers provided in the ICE list in case the first one will fall?  Do, I need to do this fallback solution manually, or will the Simple Peer library do this automatically. Thanks

**Comment by RationalCoding** [comment]
You can provide as many STUN or TURN servers as you'd like. If any are down, the fallbacks will be automatically used. STUN will be attempted first unless `iceTransportPolicy:'relay'`` is set.

Again, none of this will work behind a symmetric NAT because of the above issue.

**Comment by ibudisteanu** [comment]
@RationalCoding  I understand now. I found a new WebRTC library & demo that works on symmetric NAT + normal NAT (tested) https://demo.easyrtc.com/demos/demo_instant_messaging.html Maybe this demo will help you understand where the problem for Simple Peer because I truly believe it is a problem somewhere either in the Javascript Interface for WebRTC (Simple Peer npm) or the TURN server (less likely)

**Comment by RationalCoding** [comment]
@ibudisteanu That demo uses sendDataWS, which is over websockets, not WebRTC. It's not P2P in any sense.

**Comment by ibudisteanu** [comment]
Probably, you clicked on my first link, which I have edit it to this one https://demo.easyrtc.com/demos/demo_data_channel_messaging.html 

This demo ( that I've tested)  is using  easyrtc.sendDataP2P  and it works fine on NAT symmetric + normal NAT

**Comment by RationalCoding** [comment]
Interesting, this does indeed work without falling back to websockets. I'll see if the authors can give us any insight.

**Comment by RationalCoding** [comment]
Here's the minimal demo I've been using to reproduce if anyone is interested in helping with this issue. (You can use rabb.it to get a symmetric NAT peer).

https://codepen.io/RationalCoding/pen/bYEmjw?editors=0011

**Comment by ibudisteanu** [comment]
@RationalCoding EasyRTC looks open source  https://github.com/priologic/easyrtc

According to this [discussion](https://github.com/priologic/easyrtc/issues/19#issuecomment-28668673) on EasyRTC's GitHub, Eric is pointing out that any TURN server can be used to establish connections for symmetric NAT. This means that the problem is in Simple Peer Library, where the source code handles the communication with TURN servers.

I want to point one more interesting behavior of this issue. After the signals are successfully generated, the **error "Ice connection failed." appears in both browsers, almost simultaneously** when they want to establish the p2p WebRTC connection,

**Comment by doganmurat** [comment]
Hi all
In my case one side is on browser and the other side is on nodejs.
I have my own turn-stun server installed which is online.
l forced simple-peer to work with only relay mode then it works on symmetric-nat.
But in this mode l can not work with unsymmetric-nat.
I also think this is problem of simple-peer.

I could not find a solution yet.

**Comment by RationalCoding** [comment]
easyrtc actually throws the "NOVIABLEICE" error when in debug mode. Unfortunately it seems they have the same issue, but they fallback to websockets.

I don't know why TURN isn't able to traverse symmetric NATs, considering that's one of the reasons TURN exists at all.

My only advice is to skip TURN altogether and use a websockets fallback instead (on ice failure). You're going to be relaying data anyhow, and websockets is more reliable. If you need audio/video you may be out of luck for now.

**Comment by ibudisteanu** [comment]
@RationalCoding  I have tested the easyRTC and I don't see the any "NOVIABLEICE" error on my symmetric NAT => normal NAT communication via WebRTC protocol. Here is a screen recording that looks that the WebRTC is connecting through the TURN server

![Symmetric NAT demo](http://g.recordit.co/yDXjmceYBX.gif)

**Comment by ibudisteanu** [comment]
@doganmurat  @RationalCoding any progress so far in solving the TURN symmetric NAT issue in Simple Peer, or do I have to switch to use easyRTC?

**Comment by bparker667** [comment]
I have been using simple peer testing out an application I am working on. Everything has been working fine for a few weeks as long as I have a good turn server and credentials. Since yesterday I am getting this ICE connection failure too. Has anyone had any luck resolving this?

**Comment by artusvranken** [comment]
I've been having a similar issue, even when using the test page on your readme.md. I can start up the pages and initialize the simple-rtc peers, but when I generate an answer signal from the receiving peer and I don't paste it real quick to the initator, I get the message:

```
ICE failed, add a TURN server and see about:webrtc for more details
```

Note that I even tested it with a custom STUN and TURN server from numb.viagenie.ca but I get the same results.

**Comment by rivertam** [comment]
I've been getting this error a lot (like 20-30 times in a row), even when connection seems to succeed. It's mostly just a nuisance for me, but I'd still rather see it go away. I could definitely be doing something wrong, though. Has anyone else experienced this?

**Comment by nathansherburn** [comment]
To all the halfwits who don't read properly (of which I'm one), **make sure you paste the "answer" back into the initiator** when running the code from the README.md.

Failing to do this will result in this same error after the timeout.

**Comment by artusvranken** [comment]
@nathansherburn , no need to be rude. I do exactly that but still get the error if I don't paste it back quick enough.

**Comment by nathansherburn** [comment]
Sorry @artusvranken! Wasn't meaning to offend :) Was just trying to use a little self deprecating humour.

I spent an hour trying to debug the code before realising I just hadn't read the instructions properly and was annoyed with myself.

**Comment by AlbertoElias** [comment]
So for me it's unclear if this is an issue with TURN servers or with this library? Using the viagenie servers, I was getting the same issues except with Brave Browser. Brave worked fined with Chrome. But Chrome with Chrome or with Firefox, or Firefox with Firefox didn't work. And with Brave it worked even if I was on symmetric nat.

 It seems to work with a custom coturn server for now, still testing

**Comment by doganmurat** [comment]
Any update @RationalCoding ?
Problem still exists.
Thanks,

**Comment by doganmurat** [comment]
Hi all,

Here is my latest findings.
I was using a my own stun/turn server.
Normally stun/turn server uses 2 ports (TCP/UDP)
BUT
When one or both peer is behind a symmetric NAT, turn server opens other ports to communicate with peers.

My problem was firewall which was blocking peer ports so after that Ice connection error occurs.
Firewall or unreachable turn server could be the main reason for that error.
Could you please confirm that?

**Comment by RationalCoding** [comment]
This error is caused by either the ports needed to establish a connection to the TURN/STUN servers being blocked or the UDP protocol being blocked.

The issue with Chrome not attempting to use any TURN server appears to have been fixed.

Mobile internet providers block UDP traffic and many ports used by TURN.

Try opening all ports on your firewall and appending `?transport=tcp` to the TURN config. You can also try configuring your TURN server to use port 443 as many firewalls will treat TCP traffic on this port as opaque SSL traffic.

**Comment by ilDon** [comment]
> Ensure you can connect to your STUN/TURN servers with this tool (make sure you receive at least one relay candidate for a TURN server): https://webrtc.github.io/samples/src/content/peerconnection/trickle-ice/
> Try opening all ports on your firewall and appending ?transport=tcp to the TURN config. You can also try configuring your TURN server to use port 443 as many firewalls will treat TCP traffic on this port as opaque SSL traffic. Lastly, you can try using TLS over TCP to encrypt traffic to the TURN server which should prevent any firewall blocking.

@RationalCoding thanks! Your directions where very helpful for understanding why I was not able to connect to my mobile. I strongly suggest to add something about this network-related issue to the README, for those with less experience on webrtc it isn't obvious and it makes simple-peer look buggy, while it's an issue entirely dependant on external factors.

**Comment by rubender** [comment]
I have the same problem, to solve this I've change locally simple-peer.js 
change this 
`self._pc = new (self._wrtc.RTCPeerConnection)(self.config, self.constraints)`
to this 
```
let my_config = {
    iceTransportPolicy: 'relay',
    ...self.config,
}
self._pc = new (self._wrtc.RTCPeerConnection)(self.config, self.constraints)
```
---

problem is in pass only config of servers to RTCPeerConnection 

const peer = new SimplePeer({
      iceTransportPolicy: 'relay', // this property will ignore 
      config: {
        iceServers: [...  ]
      }
    });

**Comment by rubender** [comment]
I have the same problem, to solve this I've change locally simple-peer.js 
change this 
`self._pc = new (self._wrtc.RTCPeerConnection)(self.config, self.constraints)`
to this 
```
let my_config = {
    iceTransportPolicy: 'relay',
    ...self.config,
}
self._pc = new (self._wrtc.RTCPeerConnection)(self.config, self.constraints)
```
---

problem is in pass only config of servers to RTCPeerConnection 
```

const peer = new SimplePeer({
      iceTransportPolicy: 'relay', // this property will ignore 
      config: {
        iceServers: [...  ]
      }
    });
```

**Comment by MiguelRipoll23** [comment]
After this issue affecting me for a whole day I was able to fix this error after some researching.

1. Turn off Windows Firewall.
2. **Request video/microphone permission before starting peer connections** as Safari requires you to do this to disable ICE candidate restrictions. So **even if you don't use user media**, you must do this anyways because it means the user trusts your website. -> [Source](https://webkit.org/blog/7763/a-closer-look-into-webrtc/)

These are my ICE servers list just in case:

```
{'url': 'stun:stun.services.mozilla.com', 'urls': 'stun:stun.services.mozilla.com'},
{'url': 'stun:stun.l.google.com:19302', 'urls': 'stun:stun.l.google.com:19302'}
```

Now my local devices can connect between each other, I don't know if remote ones work yet because I recently started my project.

**Comment by artus** [comment]
Issues like this really scare away people trying to work with WebRTC. Users should not have to turn off their firewall just to use a service in their browsers.

**Comment by rivertam** [comment]
@artus Agreed, but there's nothing this repository can do about it.

**Comment by MiguelRipoll23** [comment]
My remote connections don’t work, something that I did expect to be honest. I don’t understand why WebRTC doesn’t try to at least open a port on the network to allow communication, this is something that every videogame thst uses voice chat does.

WebRTC is a complex but ineffective protocol.

**Comment by tadejrola** [comment]
Worked well on one machine, different browsers, didnt work on same/different network, different machines.

After **turning off** firewall on both machines, it works. This is not how it is supposed to be done. What else do you recommend? Tried TURN/STUN servers, did not work without turning off firewall.

**Comment by t-mullen** [comment]
It depends on the configuration of your firewall. You could have set your firewall ruleset to block all traffic completely, or blocked the ports your STUN/TURN servers are on, or blocked UDP traffic. It's impossible to help debug firewall configurations without knowing what they are.

I've pinned my troubleshooting steps to this issue, please follow them and open a **new issue** with a filtered pcap file if they do not help.

**Comment by vamshi9666** [comment]
any updates regarding this ?

**Comment by t-mullen** [comment]
Review the pinned troubleshooting and open a specific issue if those steps do not work for you.

**Comment by yaniv-fink** [comment]
It seems that this problem was not solved yet.
I have the same problem, getting Ice connection failed 
Only when creating video from firefox to chrome (I am using simple peer and Angular 7)
My observation is this.

This is not a firewall problem. (happen in my network, and I closed all firewall and antivirus 
This is not STUN TURN problem, I checked the URLs they are all valid, and I used big companies like Twilio and Xirsys both lead to the same error

It might be some chrome timeout after 15 seconds, or problem in the simple peer library, I am not sure 
I noticed that someone opened a chrome bug
https://bugs.chromium.org/p/chromium/issues/detail?id=811000
But I checked in later chrome versions the bug was not solved.

Here is some data while debugging the WebRTC
![image](https://user-images.githubusercontent.com/9554134/50828917-7224cc80-134b-11e9-8e8b-3ae860337544.png)

**Comment by yaniv-fink** [comment]
It seems that the problem was when I loaded the IceServer from my server, into the peer object I used a wrong configuration, (did not initiate it properly) 
now it is working, (although it is not working from chrome using iPhone, I guess this is Apple lack of support in WebRTC, but this is another story...)
